geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/function/BinaryOutputEncoder.scala

```diff
@@ -0,0 +1,226 @@
+    validateLatLong(latLon, sft)
+
+    val sysTime = System.currentTimeMillis
+    val dtgIndex = sft.indexOf(dtgField)
+
+    // get date function based
+    val getDtg: (SimpleFeature) => Long = (f) => {
+      val date = f.getAttribute(dtgIndex).asInstanceOf[Date]
+      if (date == null) sysTime else date.getTime
+    }
+    // for line strings, we need an array of dates corresponding to the points in the line
+    val getLineDtg: (SimpleFeature) => Array[Long] = (f) => {
+      val dates = f.getAttribute(dtgIndex).asInstanceOf[java.util.List[Date]]
+      if (dates == null) Array.empty else dates.map(_.getTime).toArray
+    }
+
+    // get lat/lon as floats
+    val getLatLon: (SimpleFeature) => (Float, Float) =
+      latLon.map { case (lat, lon) =>
+        // if requested, export arbitrary fields as lat/lon
+        val latIndex = sft.indexOf(lat)
+        val lonIndex = sft.indexOf(lon)
+        val binding = sft.getType(latIndex).getBinding
+        if (binding == classOf[java.lang.Float]) {
+          (f: SimpleFeature) =>
+            (f.getAttribute(latIndex).asInstanceOf[Float], f.getAttribute(lonIndex).asInstanceOf[Float])
+        } else if (binding == classOf[java.lang.Double]) {
+          (f: SimpleFeature) =>
+            (f.getAttribute(latIndex).asInstanceOf[Double].toFloat, f.getAttribute(lonIndex).asInstanceOf[Double].toFloat)
+        } else {
+          (f: SimpleFeature) =>
+            (f.getAttribute(latIndex).toString.toFloat, f.getAttribute(lonIndex).toString.toFloat)
+        }
+      }.getOrElse {
+        // default is to use the geometry
+        // depending on srs requested and wfs versions, axis order can be flipped
+        axisOrder match {
+          case LatLon => (f) => {
+            val geom = f.getDefaultGeometry.asInstanceOf[Geometry].getInteriorPoint
+            (geom.getX.toFloat, geom.getY.toFloat)
+          }
+          case LonLat => (f) => {
+            val geom = f.getDefaultGeometry.asInstanceOf[Geometry].getInteriorPoint
+            (geom.getY.toFloat, geom.getX.toFloat)
+          }
+        }
+      }
+    // for linestrings, we return each point - use an array so we get constant-time lookup
+    // depending on srs requested and wfs versions, axis order can be flipped
+    val getLineLatLon: (SimpleFeature) => Array[(Float, Float)] = axisOrder match {
+      case LatLon => (f) => {
+        val geom = f.getDefaultGeometry.asInstanceOf[LineString]
+        val points = (0 until geom.getNumPoints).map(geom.getPointN(_))
+        points.map(p => (p.getX.toFloat, p.getY.toFloat)).toArray
+      }
+      case LonLat => (f) => {
+        val geom = f.getDefaultGeometry.asInstanceOf[LineString]
+        val points = (0 until geom.getNumPoints).map(geom.getPointN(_))
+        points.map(p => (p.getY.toFloat, p.getX.toFloat)).toArray
+      }
+    }
+
+    // gets the track id from a feature
+    val getTrackId: (SimpleFeature) => Option[String] = trackIdField match {
```

Comment:
ah nvm these are matches that generate functions. You're good. I think a match statement actually creates an anonymous partial function just to let you know. So if you do i match { case i: Int i + 1} what you are actually doing is val anonoymousFunction = PartialFunction[Int, Int] { case i: Int => i + }; anonoymousFunction(i)}